curp/src/server/raw_curp/log.rs

```diff
@@ -43,14 +43,135 @@ pub(super) struct Log<C: Command> {
     pub(super) last_exe: LogIndex,
     /// Tx to send log entries to persist task
     log_tx: mpsc::UnboundedSender<LogEntry<C>>,
-    /// Prefix sum vector of log entries
-    batch_index: VecDeque<u64>,
     /// Batch size limit
     batch_limit: u64,
     /// Entries to keep in memory
     entries_cap: usize,
 }
 
+/// That's a struct to store log entries and calculate batch of log
+#[derive(Debug)]
+struct LogEntryVecDeque<C: Command> {
+    /// A VecDeque to store log entries, it will be serialized and persisted
+    entries: VecDeque<LogEntry<C>>,
+    /// The sum of serialized size of previous log entries
+    /// batch_index[i+1] = batch_index[i] + size(entries[i])
+    batch_index: VecDeque<u64>,
+}
+
+impl<C: Command> LogEntryVecDeque<C> {
+    /// return a log entries with cap
+    fn new(cap: usize) -> Self {
+        let mut batch_index = VecDeque::with_capacity(cap);
+        batch_index.push_back(0);
+        Self {
+            /// Now keep the old logic to init entries with LOG_INIT_CAPACITY
+            /// TODO: maybe use the cap in args to init vec?
+            /// - entries: cap ?
+            /// - batch_index: cap + 1 ?
+            entries: VecDeque::with_capacity(LOG_INIT_CAPACITY),
+            batch_index,
+        }
+    }
+
+    /// Shortens the log entries, keeping the first `len` elements and dropping
+    /// the rest.
+    /// `batch_index` will keep len+1 elem
+    fn truncate(&mut self, len: usize) {
+        self.entries.truncate(len);
+        self.batch_index.truncate(len.overflow_add(1));
+    }
+
+    /// push a log entry into the back of queue
+    fn push_back(&mut self, entry: LogEntry<C>) -> Result<(), bincode::Error> {
+        let entry_size = serialized_size(&entry)?;
+
+        self.entries.push_back(entry);
+        let pre_entry_size = if let Some(&last_entry_size) = self.batch_index.back() {
+            last_entry_size
+        } else {
+            0
+        };
```

Comment:
Return an error when `batch_index.back()` is `None`? it should always be `Some`

Reply:
I think it's not reasonable to throw an error, how about `panic!` at `else` branch? It must be a programmatic bug if `back()` is `None`.

Reply:
How about `let pre_entries_size = self.batch_index.back().unwrap_or_else(|| unreachable!("batch_index.back() should never be None"));`

Reply:
How about `let pre_entries_size = self.batch_index.back().unwrap_or_else(|| unreachable!("batch_index.back() should never be None"));`